doc/cheops.6
.TH CHEOPS 6 \" -*- nroff -*-
.SH NAME
CHEOPS \- Chess Opponent Simulator
.SH SYNOPSIS
  cheops [\-\-help] [\-\-version]
.SH DESCRIPTION
.P
Cheops (CHEss OPponent Simulator) is a fully-functional chess program
capable of human-vs-human, human-vs-computer, and computer-vs-computer
play.
.SH OPTIONS
.P
.TP
.BI "\-\-help" ""
Display a short help message and exit.
.TP
.BI "\-\-version" ""
Display version information and exit.
.SH GAMEPLAY
Once invoked, CHEOPS displays a title screen and prompts the user for
who controls each player. If a computer opponent is chosen, a list of
default static evaluation function weights is displayed.  To accept
the default values, type \fBdone\fR; otherwise, enter the number of
the weight to change. The program will then prompt for the new value.
The program will allow the user to adjust the weights until \fBdone\fR
is entered.
.P
Following player setup, the board is displayed and the game begins.
The board is represented by a simple text diagram using the standard
one-character piece abbreviations; white pieces are represented by
small letters and black pieces by capitals.  Each player makes his
move in turn; human players are prompted for their move, and computer
players move automatically. To make a move, enter the source and
destination coordinates in standard algebraic notation -- for example,
\fBg1f3\fR moves the knight at g1 to f3.  For pawn promotions, the
promotion piece must be appended -- for instance, \fBe7e8n\fR promotes
to a knight. For castling, the user may either enter the source and
destination coordinates of the king, or he may simply type \fBo-o\fR
or \fBo-o-o\fR for kingside or queenside castling, respectively. The
players may continue entering moves in this manner until a checkmate
or draw state is reached. As in regular chess, a draw occurs upon
stalemate, triple observance of the same board configuration with the
same side to move, or observance of fifty moves without a capture or
pawn advance.
.P
In addition to legal moves, the following commands are also available
as responses for human players. At the end of a game, no further moves
are possible, and one of these commands must be entered.
.TP
.BI "help" ""
display command summary
.TP
.BI "resign" ""
forfeit the match
.TP
.BI "bd" ""
redisplay the board
.TP
.BI "coords" ""
toggle display of board coordinates
.TP
.BI "reverse" ""
reverse board display (black on top to white on top, or vice versa)
.TP
.BI "think" ""
toggle display of computer ``thinking''
.TP
.BI "log" ""
display log of moves for the current game
.TP
.BI "save" ""
save the move log to a file
.TP
.BI "new" ""
start a new game (board in initial configuration with white to move)
.TP
.BI "who" ""
set players and computer stats, if applicable
.TP
.BI "quit" ""
exit the program
.SH BUGS
Report bugs to <https://github.com/logological/cheops/issues>.
.SH SEE ALSO
gnuchess(6), chesstool(6), xboard(6)
.P
CHEOPS home page: http://www.nothingisreal.com/cheops/
.SH AUTHOR
Tristan Miller <psychonaut@nothingisreal.com>.
.SH COPYRIGHT
Copyright (C) 1999-2016 Tristan Miller.
.P
Permission is granted to make and distribute verbatim or modified
copies of this manual provided the copyright notice and this
permission notice are preserved on all copies.


NEWS
CHEOPS NEWS -- history of user-visible changes
$Date$

Copyright (C) 2015 Tristan Miller.  See the end of this file for
copying conditions.

Please send CHEOPS bug reports to Tristan Miller
<psychonaut@nothingisreal.com>.


Version 1.2 (2015-02-01)

* Fixed some problems which prevented compilation with modern
  (and more standards-compliant) versions of GCC

Version 1.1 (2003-12-08)

* CHEOPS is now packaged in a manner similar to most GNU programs,
  including a `configure' configuration/installation script suitable
  for Unix-like systems.
* The LaTeX manual has been revised.
* A Unix man page is now available.
* The --version and --help command-line options are now supported.

Version 1.0 (2003-11-23)

* Initial release

----------------------------------------------------------------------
Copyright information:

Copyright (C) 2015 Tristan Miller.

   Permission is granted to anyone to make or distribute verbatim
   copies of this document as received, in any medium, provided that
   the copyright notice and this permission notice are preserved, thus
   giving the recipient permission to redistribute in turn.

   Permission is granted to distribute modified versions of this
   document, or of portions of it, under the above conditions,
   provided also that they carry prominent notices stating who last
   changed them.


src/cheops.cpp
//
// cheops.cpp : Defines the entry point for the console application.
//
// Copyright (C) 2003 Tristan Miller <psychonaut@nothingisreal.com>
//  
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as
// published by the Free Software Foundation; either version 2 of the
// License, or (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
// 02111-1307, USA.
//
//////////////////////////////////////////////////////////////////////

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include "stdafx.h"
#include "ChessBoard.h"
#include "HumanPlayer.h"
#include "ComputerPlayer.h"
#include <cstdio>
#include <cstring>

void show_menu(void);
chessmove *parse_move(const string& s);
void who_plays(Player *&w, Player *&b);
void show_help(void);
void show_version(void);

int main(int argc, char* argv[]) {

    ChessBoard c;
    Player *w_player=NULL, *b_player=NULL, *this_player;
    string cmd, log;
    game_status status=Normal;
    chessmove *m = NULL;
    int turn=0;

    // Process command-line parameters
    while (argc-- > 1) {
      argv++;
      if (strcmp(*argv, "--help") == 0) {
	show_help();
	exit(0);
      }
      else if (strcmp(*argv, "--version") == 0) {
	show_version();
	exit(0);
      }
      else {
	cerr << "cheops: unrecognized option `" << argv[0] << "'\n";
	show_help();
	exit(1);
      }
    }

    // Print title screen
    show_version();
    cout <<
      "\t\t\t__   __\n"
      "\t\t\t\\ \\ / /\n"
      "\t\t\t \\ v / ___ ___  _  _  _\n"
      "\t\t\t  > < / __) _ \\| || || |\n"
      "\t\t\t / ^ \\> _| (_) ) \\| |/ |\n"
      "\t\t\t/_/ \\_\\___)___/ \\_   _/\n"
      "\t\t\t _________________| |__\n"
      "\t\t\t(_________________|_|__)\n\n"
      ;

    who_plays(w_player, b_player);

    cout << "\nEnter \"help\" or \"?\" for the menu.\n" << c;

    // Main loop
    while (1) {
        cout.flush();
        this_player=c.white_moves()?w_player:b_player;
        cmd=this_player->get_command(c);

        if (cmd=="board" || cmd=="bd")
            cout << c;
        else if (cmd=="reverse")
            c.show_reverse=!c.show_reverse;
        else if (cmd=="coords")
            c.show_coordinates=!c.show_coordinates;
        else if (cmd=="think")
            ComputerPlayer::show_thinking=!ComputerPlayer::show_thinking;
        else if (cmd=="help" || cmd=="?")
            show_menu();
        else if (cmd=="who")
            who_plays(w_player, b_player);
        else if (cmd=="save") {
            ofstream logfile("cheops.log");
            logfile << log << endl;
            logfile.close();
        }
        else if (cmd=="new") {
            c=ChessBoard();
            turn=0;
            log=string();
        }
        else if (cmd=="quit" || cmd=="exit")
            break;
        else if (cmd=="log")
            cout << log << endl;
        else if (cmd=="resign") {
            cout << "You resign. " << (c.white_moves()?"Black":"White") << 
                " wins!\n";
            c.in_progress=false;
        }
        else {
            delete m;
            // Kingside castling
            if (cmd=="o-o") {
                m = new chessmove;
                m->promotion=Empty;
                m->from=this_player==w_player?e1:e8;
                m->to=this_player==w_player?g1:g8;
            }
            // Queenside castling
            else if (cmd=="o-o-o") {
                m = new chessmove;
                m->promotion=Empty;
                m->from=this_player==w_player?e1:e8;
                m->to=this_player==w_player?c1:c8;
            }
            // Regular move
            else
                m=parse_move(cmd);
            if (c.in_progress && (!this_player->human || m && c.can_move(*m))) {
                status=c.do_move(*m);
                cout << c;

                // Update board log
                if (turn%2==0) {
                    char tmp[6];
                    sprintf(tmp,"%2d. ",turn/2+1);
                    log+=tmp;
                    log+=*m;
                }
                else {
                    log+='\t';
                    log+=*m;
                    log+='\n';
                }
                turn++;

                // Display board status
                switch (status) {
                case Checkmate:
                    cout << "Checkmate!\n";
                    break;
                case Stalemate:
                    cout << "Stalemate!\n";
                    break;
                case TripleOccurrence:
                    cout << "Triple occurrence -- game is drawn.\n";
                    break;
                case FiftyMoves:
                    cout << "Fifty move rule -- game is drawn.\n";
                case Check:
                    cout << "Check!\n";
                    break;
                }
            }
            else
                cout << "Illegal move\n";
        }
    }

    return 0;
}

// -------------------------------------------------------------------
// Display the menu
// -------------------------------------------------------------------
void show_menu(void) {
    cout << "\nCHEOPS command summary\n";
    cout << "---------------------------------------------------------\n";
    cout << "a1b3\tmove from a1 to b3\n";
    cout << "a7a8n\tpromote pawn to knight\n";
    cout << "o-o\tcastle king-side\n";
    cout << "o-o-o\tcastle queen-side\n";
//  cout << "draw\toffer/force draw\n";
    cout << "resign\tresign\n";
    cout << "bd\tredisplay board\n";
    cout << "coords\ttoggle coordinates\n";
    cout << "reverse\treverse board display\n";
    cout << "think\ttoggle display of computer thinking\n";
    cout << "log\tlog of moves\n";
    cout << "save\tsave log to cheops.log\n";
    cout << "new\tnew game\n";
    cout << "who\twho plays (modify current players)\n";
    cout << "quit\tquit program\n";
    cout << "---------------------------------------------------------\n";
}

// -------------------------------------------------------------------
// Parse move
//  - returns pointer to a chessmove structure on success, else NULL
// -------------------------------------------------------------------
chessmove *parse_move(const string& s) {

    // Is this a valid move string?
    int len=s.length();
    if (len>5 || len<4 || s[0]<'a' || s[0]>'h' || s[2]<'a' ||
        s[2]>'h' || s[1]<'1' || s[1]>'8' || s[3]<'1' || s[3]>'8')
        return NULL;
    if (len==5 && !strchr("qrnb", s[4]))
        return NULL;

    // Determine promotion type
    chessmove *m = new chessmove;
    if (len==5) {
        switch(s[4]) {
        case 'q':
            m->promotion=Queen;
            break;
        case 'r':
            m->promotion=Rook;
            break;
        case 'n':
            m->promotion=Knight;
            break;
        case 'b':
            m->promotion=Bishop;
            break;
        }
    }
    else 
        m->promotion=Empty;

    // Determine source and destination squares
    m->from=s[0]-'a'+(s[1]-'1')*8;
    m->to=s[2]-'a'+(s[3]-'1')*8;

    return m;
}

// -------------------------------------------------------------------
// Set players
// -------------------------------------------------------------------
void who_plays(Player *&w, Player *&b) {
    string s;
    while (s!="c" && s!="h") {
        cout << "White [c]omputer or [h]uman? ";
        cin >> s;
        cout << endl;
    }
    if (s=="h") {
        if (!(w && w->human==true)) {
            delete w;
            w = new HumanPlayer;
        }
    }
    else {
        if (!(w && w->human==false)) {
            delete w;
            w = new ComputerPlayer;
        }
        else
            w->set_stats();
    }

    s=string();
    while (s!="c" && s!="h") {
        cout << "Black [c]omputer or [h]uman? ";
        cin >> s;
        cout << endl;
    }
    if (s=="h") {
        if (!(b && b->human==true)) {
            delete b;
            b = new HumanPlayer;
        }
    }
    else {
        if (!(b && b->human==false)) {
            delete b;
            b = new ComputerPlayer;
        }
        else
            b->set_stats();
    }
}

void show_help(void) {
  cerr <<
    "Usage: cheops [options]\n"
    "Options:\n"
    "--help     Display this information and exit\n"
    "--version  Display version information and exit\n"
    ;
}

void show_version(void) {
  cout <<
    PACKAGE_STRING "\n"
    "Copyright (C) 1999-2016 Tristan Miller\n"
    "This is free software; see the source for copying conditions.  There is NO\n"
    "warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n"
    ;
}
